### DISCUSSION TITLE
A simple playlist and icecast2

### DISCUSSION THREAD

**Discussion by newmediacrew** [savonet/liquidsoap#2569] [Q&A]
So I am trying to setup liquidsoap and icecast2 to start streaming some music. I had to switch to liquid soap cause ezstream couldn't play mp3 files cause of a bug. https://github.com/xiph/ezstream/issues/30#issuecomment-1220979847

For some reason I am not able to "login" to the ice cast server. Probably due to the lack of configuration settings.

**icecast.xml**
<icecast>
    
    <location>Earth</location>
    <admin>icemaster@localhost</admin>

    

    <limits>
        <clients>100</clients>
        <sources>2</sources>
        <queue-size>524288</queue-size>
        <client-timeout>30</client-timeout>
        <header-timeout>15</header-timeout>
        <source-timeout>10</source-timeout>
        
        <burst-on-connect>1</burst-on-connect>
        
        <burst-size>65535</burst-size>
    </limits>

    <authentication>
        
        <source-password>supersafepassword</source-password>
        
        <relay-password>supersafepassword</relay-password>

        
        <admin-user>admin</admin-user>
        <admin-password>supersafepassword</admin-password>
    </authentication>

    

    
    

    
    <hostname>localhost</hostname>

    
    <listen-socket>
        <port>8000</port>
        
        
    </listen-socket>
    
    

    
    <http-headers>
        <header name="Access-Control-Allow-Origin" value="*" />
    </http-headers>

    
    
    
    
    

    
    

    

    

    
    

    
    

    

    <fileserve>1</fileserve>

    <paths>
        
        <basedir>/usr/share/icecast2</basedir>

        
        <logdir>/var/log/icecast2</logdir>
        <webroot>/usr/share/icecast2/web</webroot>
        <adminroot>/usr/share/icecast2/admin</adminroot>
        

        
        
        
        <alias source="/" destination="/status.xsl"/>
        
    </paths>

    <logging>
        <accesslog>access.log</accesslog>
        <errorlog>error.log</errorlog>
        
        <loglevel>3</loglevel> 
        <logsize>10000</logsize> 
        
        
    </logging>

    <security>
        <chroot>0</chroot>
        
    </security>
</icecast>

**The liquidsoap bash script**
```
#!/usr/bin/liquidsoap
set("init.allow_root",true)
output.icecast(%mp3, host = "localhost", port = 8000, password = "supersafepassword", mount = "liq.ogg", mksafe(playlist("/root/music/loungeMusic/loungeMusic.m3u")))
```

**The log**
**root@southjanus:~/music# ./liquidSoap.sh 
2022/08/19 21:31:25 >>> LOG START
2022/08/19 21:31:24 [main:3] Liquidsoap 2.0.2
2022/08/19 21:31:24 [main:3] Using: graphics=5.1.2 bytes=[distributed with OCaml 4.02 or above] pcre=7.5.0 sedlex=2.5 menhirLib=20210929 dtools=0.4.4 duppy=0.9.2 cry=0.6.5 mm=0.7.3 xmlplaylist=0.1.5 ogg=0.7.0 ogg.decoder=0.7.0 vorbis=0.8.0 vorbis.decoder=0.8.0 opus=0.2.1 opus.decoder=0.2.1 speex=0.4.1 speex.decoder=0.4.1 mad=0.5.0 flac=0.3.0 flac.ogg=0.3.0 flac.decoder=0.3.0 dynlink=[distributed with Ocaml] lame=0.3.5 shine=0.2.2 gstreamer=0.3.1 frei0r=0.1.2 theora=0.4.0 theora.decoder=0.4.0 ffmpeg=1.1.1 bjack=0.1.6 alsa=0.3.0 ao=0.2.3 samplerate=0.1.6 taglib=0.3.7 ssl=0.5.9 magic=0.7.3 camomile=1.0.2 inotify=2.3 yojson=[unspecified] faad=0.5.0 soundtouch=0.1.9 portaudio=0.2.2 pulseaudio=0.1.4 ladspa=0.2.0 camlimages=4.2.6 lo=0.2.0 gd=1.0a5
2022/08/19 21:31:24 [dynamic.loader:3] Could not find dynamic module for fdkaac encoder.
2022/08/19 21:31:24 [gstreamer.loader:3] Loaded GStreamer 1.20.3 0
2022/08/19 21:31:24 [clock:3] Using builtin (low-precision) implementation for latency control
2022/08/19 21:31:25 [lang:2] WARNING: "set" is deprecated and will be removed in future version. Please use `settings.path.to.key.set(value)`
2022/08/19 21:31:25 [frame:3] Using 44100Hz audio, 25Hz video, 44100Hz main.
2022/08/19 21:31:25 [frame:3] Video frame size set to: 1280x720
2022/08/19 21:31:25 [frame:3] Frame size must be a multiple of 1764 ticks = 1764 audio samples = 1 video samples.
2022/08/19 21:31:25 [frame:3] Targeting 'frame.duration': 0.04s = 1764 audio samples = 1764 ticks.
2022/08/19 21:31:25 [frame:3] Frames last 0.04s = 1764 audio samples = 1 video samples = 1764 ticks.
2022/08/19 21:31:25 [sandbox:3] Could not find binary bwrap, disabling sandboxing.
2022/08/19 21:31:25 [video.converter:3] Using preferred video converter: ffmpeg.
2022/08/19 21:31:25 [audio.converter:3] Using samplerate converter: ffmpeg.
2022/08/19 21:31:25 [liq.ogg:3] Connecting mount liq.ogg for source@localhost...
2022/08/19 21:31:25 [liq.ogg:2] Connection failed: 401, Authentication Required (HTTP/1.0)
2022/08/19 21:31:25 [liq.ogg:3] Will try again in 3.00 sec.
2022/08/19 21:31:25 [clock.main:3] Streaming loop starts in auto-sync mode
2022/08/19 21:31:25 [clock.main:3] Delegating synchronisation to CPU clock
2022/08/19 21:31:25 [mksafe:3] Switch to safe_blank.
2022/08/19 21:31:25 [loungeMusic.m3u:3] Prepared "/root/music/loungeMusic/4l5nfBzl19g.mp3" (RID 1).
2022/08/19 21:31:25 [mksafe:3] Switch to loungeMusic.m3u with transition.
[mp3float @ 0x7f9988020c80] Could not update timestamps for skipped samples.
2022/08/19 21:31:29 [liq.ogg:3] Connecting mount liq.ogg for source@localhost...
2022/08/19 21:31:29 [liq.ogg:2] Connection failed: 401, Authentication Required (HTTP/1.0)
2022/08/19 21:31:29 [liq.ogg:3] Will try again in 3.00 sec.
^C2022/08/19 21:31:29 [main:3] Shutdown started!**

What am I missing here ?

Thank you for the help

**Comment by vitoyucepi** [comment]
Hi, @newmediacrew.

Here's a simple config.
```xml
<?xml version="1.0"?>
<icecast>
  <location>Earth</location>
  <admin>icemaster@localhost</admin>
  <limits>
    <clients>100</clients>
    <sources>2</sources>
    <queue-size>524288</queue-size>
    <client-timeout>30</client-timeout>
    <header-timeout>15</header-timeout>
    <source-timeout>10</source-timeout>
    <burst-on-connect>1</burst-on-connect>
    <burst-size>65535</burst-size>
  </limits>
  <authentication>
    <source-password>hackme</source-password>
    <relay-password/>
    <admin-user>admin</admin-user>
    <admin-password/>
  </authentication>
  <hostname>localhost</hostname>
  <listen-socket>
    <port>8000</port>
  </listen-socket>
  <http-headers>
    <header name="Access-Control-Allow-Origin" value="*"/>
  </http-headers>
  <mount type="default">
    <charset>UTF-8</charset>
  </mount>
  <fileserve>1</fileserve>
  <paths>
    <basedir>/usr/share/icecast</basedir>
    <logdir>/var/log/icecast</logdir>
    <webroot>/usr/share/icecast/web</webroot>
    <adminroot>/usr/share/icecast/admin</adminroot>
    <pidfile/>
    <alias source="/" destination="/status.xsl"/>
  </paths>
  <logging>
    <accesslog>-</accesslog>
    <errorlog>-</errorlog>
    <playlistlog/>
    <loglevel>3</loglevel>
    <logsize>0</logsize>
    <logarchive>0</logarchive>
  </logging>
  <security>
    <chroot>0</chroot>
    <changeowner>
      <user>icecast</user>
      <group>icecast</group>
    </changeowner>
  </security>
</icecast>
```
Let's say icecast server runs on the hostname `icecast`.
If I start icecast with this config, then I can `curl http://icecast:8000`.
It's possible to stream audio with command `ffmpeg -re -i example.ogg -vn -c:a copy -f ogg icecast://source:hackme@localhost:8000/radio`.
And listen with `ffplay http://localhost:8000/radio`.

I set up port to listen here
```xml
<listen-socket>
  <port>8000</port>
</listen-socket>
```
I think it's possible to use another streaming user, but `source` is fine.
I use password `hackme`, for testing purpose
```xml
<authentication>
  <source-password>hackme</source-password>
</authentication>
```
That's all for icecast.

Liquidsoap config
```ruby
radio = single("/var/liquidsoap/example.ogg")

output.icecast(
  host = 'icecast',
  port = 8000,
  user = 'source',
  password = 'hackme',
  mount = 'radio',
  %vorbis,
  radio
)
```

I will stream to host `icecast:8000` with user `source` and password `hackme`, the mount is `/radio`.
Like in ffmpeg `icecast://source:hackme@icecast:8000/radio`.
